Mod paths with `[` and `]` cause loading to fail

As reported in discord.

This results in an error like this: 
```
java.lang.RuntimeException: Could not execute entrypoint stage 'client' due to errors, provided by 'inventoryhud'!
	at Not Enough Crashes deobfuscated stack trace.(1.18.1+build.22)
	at org.quiltmc.loader.impl.entrypoint.EntrypointUtils.lambda$invoke0$2(EntrypointUtils.java:63)
	at org.quiltmc.loader.impl.util.ExceptionUtil.gatherExceptions(ExceptionUtil.java:33)
	at org.quiltmc.loader.impl.entrypoint.EntrypointUtils.invoke0(EntrypointUtils.java:61)
	at org.quiltmc.loader.impl.entrypoint.EntrypointUtils.invokeContainer(EntrypointUtils.java:46)
	at org.quiltmc.loader.impl.entrypoint.EntrypointUtils.invoke(EntrypointUtils.java:33)
	at org.quiltmc.loader.impl.game.minecraft.Hooks.startClient(Hooks.java:53)
	at net.minecraft.client.MinecraftClient.<init>(MinecraftClient:443)
	at net.minecraft.client.main.Main.main(Main:198)
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:568)
	at org.quiltmc.loader.impl.game.minecraft.MinecraftGameProvider.launch(MinecraftGameProvider.java:423)
	at org.quiltmc.loader.impl.launch.knot.Knot.launch(Knot.java:76)
	at org.quiltmc.loader.impl.launch.knot.KnotClient.main(KnotClient.java:24)
Caused by: org.quiltmc.loader.impl.entrypoint.QuiltEntrypointException: Exception while loading entries for entrypoint 'client' provided by 'inventoryhud'
	at org.quiltmc.loader.impl.entrypoint.EntrypointStorage.lambda$getEntrypointContainers$1(EntrypointStorage.java:222)
	at org.quiltmc.loader.impl.entrypoint.EntrypointContainerImpl.getEntrypoint(EntrypointContainerImpl.java:51)
	at org.quiltmc.loader.impl.entrypoint.EntrypointUtils.lambda$invoke$0(EntrypointUtils.java:33)
	at org.quiltmc.loader.impl.entrypoint.EntrypointUtils.invoke0(EntrypointUtils.java:59)
	... 12 more
Caused by: org.quiltmc.loader.impl.util.ExceptionUtil$WrappedException: java.net.URISyntaxException: Illegal character in path at index 75: file:/C:/Users/mDragon/AppData/Roaming/.minecraft/mods/InventoryHUD-fabric-[1.19]+-3.4.2.jar
	at org.quiltmc.loader.impl.util.ExceptionUtil.wrap(ExceptionUtil.java:51)
	at org.quiltmc.loader.impl.util.UrlUtil.asPath(UrlUtil.java:67)
	at org.quiltmc.loader.impl.launch.knot.KnotClassDelegate.lambda$getMetadata$0(KnotClassDelegate.java:196)
	at java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1708)
	at org.quiltmc.loader.impl.launch.knot.KnotClassDelegate.getMetadata(KnotClassDelegate.java:190)
	at org.quiltmc.loader.impl.launch.knot.KnotClassDelegate.getMetadata(KnotClassDelegate.java:186)
	at org.quiltmc.loader.impl.launch.knot.KnotClassDelegate.tryLoadClass(KnotClassDelegate.java:153)
	at org.quiltmc.loader.impl.launch.knot.KnotClassLoader.loadClass(KnotClassLoader.java:156)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:520)
	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(Class.java:467)
	at org.quiltmc.loader.impl.util.DefaultLanguageAdapter.create(DefaultLanguageAdapter.java:49)
	at org.quiltmc.loader.impl.entrypoint.EntrypointStorage$NewEntry.getOrCreate(EntrypointStorage.java:112)
	at org.quiltmc.loader.impl.entrypoint.EntrypointStorage.lambda$getEntrypointContainers$1(EntrypointStorage.java:220)
	... 15 more
Caused by: java.net.URISyntaxException: Illegal character in path at index 75: file:/C:/Users/mDragon/AppData/Roaming/.minecraft/mods/InventoryHUD-fabric-[1.19]+-3.4.2.jar
	at java.net.URI$Parser.fail(URI.java:2974)
	at java.net.URI$Parser.checkChars(URI.java:3145)
	at java.net.URI$Parser.parseHierarchical(URI.java:3227)
	at java.net.URI$Parser.parse(URI.java:3175)
	at java.net.URI.<init>(URI.java:623)
	at java.net.URL.toURI(URL.java:1056)
	at org.quiltmc.loader.impl.util.UrlUtil.asPath(UrlUtil.java:65)
	... 27 more
```

A likely culprit is `UrlUtil.asPath`, which might need to perform url unencoding

--- a/src/main/java/org/quiltmc/loader/impl/util/UrlUtil.java
+++ b/src/main/java/org/quiltmc/loader/impl/util/UrlUtil.java
@@ -64,15 +64,15 @@ public final class UrlUtil {
 	public static Path asPath(URL url) {
 		try {
 			return Paths.get(
-					new URI(
-							url.getProtocol(),
-							url.getUserInfo(),
-							url.getHost(),
-							url.getPort(),
-							url.getPath(),
-							url.getQuery(),
-							null
-					)
+				new URI(
+					url.getProtocol(),
+					url.getUserInfo(),
+					url.getHost(),
+					url.getPort(),
+					url.getPath(),
+					url.getQuery(),
+					null
+				)
 			);
 		} catch (URISyntaxException e) {
 			throw ExceptionUtil.wrap(e);

--- a/src/main/java/org/quiltmc/loader/impl/util/UrlUtil.java
+++ b/src/main/java/org/quiltmc/loader/impl/util/UrlUtil.java
@@ -19,6 +19,7 @@ package org.quiltmc.loader.impl.util;
 import java.io.File;
 import java.net.JarURLConnection;
 import java.net.MalformedURLException;
+import java.net.URI;
 import java.net.URISyntaxException;
 import java.net.URL;
 import java.net.URLConnection;
@@ -62,7 +63,17 @@ public final class UrlUtil {
 
 	public static Path asPath(URL url) {
 		try {
-			return Paths.get(url.toURI());
+			return Paths.get(
+					new URI(
+							url.getProtocol(),
+							url.getUserInfo(),
+							url.getHost(),
+							url.getPort(),
+							url.getPath(),
+							url.getQuery(),
+							null
+					)
+			);
 		} catch (URISyntaxException e) {
 			throw ExceptionUtil.wrap(e);
 		}

